dialog/file_xdg_test.go

```diff
@@ -0,0 +1,44 @@
+// +build linux openbsd freebsd netbsd
+// +build !android
+
+package dialog
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFavoriteLocations(t *testing.T) {
+	favoriteLocations, _ := getFavoriteLocations()
+
+	// This is a map containing the base name
+	// of every key that should be in favoriteLocations
+	expected := map[string]string{
+		"Home":      "",
+		"Documents": "Documents",
+		"Downloads": "Downloads",
+	}
+
+	homeDir, err := os.UserHomeDir()
+	assert.Nil(t, err)
+
+	for e := range expected {
+		fav, ok := favoriteLocations[e]
+		if assert.True(t, ok) {
+			if expected[e] == "" {
+				assert.Equal(t, homeDir, fav)
+			} else {
+				assert.Equal(t, homeDir, filepath.Dir(fav))
+				if _, err := os.Stat(fav); os.IsNotExist(err) {
```

Comment:
I don't understand this test sorry.
It seems to iterate expected directories and confirms that they end with the expected string.
But the XDG code exists to make this configurable - so as soon as someone has that set these will fail won't they?

The purpose of testing the favourite location lookup would be to test the wrapper around xdg-user-dir I would think, which is based on environment variables, so can we manipualte them to test the code is working?

Reply:
> The way it works is that xdg-user-dirs-update is run very early in the login phase. This program reads a configuration file, and a set of default directories. It then creates localized versions of these directories in the users home directory and sets up a config file in $(XDG_CONFIG_HOME)/user-dirs.dirs (XDG_CONFIG_HOME defaults to ~/.config) that applications can read to find these directories. 

This is from [here](https://freedesktop.org/wiki/Software/xdg-user-dirs/).

So it's not really about environment variables, but more about a config file. I think this is too "hacky" for me so I will leave it at that.

> so as soon as someone has that set these will fail won't they?

The paths are only getting checked against those strings when the folders don't exist. So this test is not testing wether ```getFavoriteLocations``` returns the correct paths. It is only checking if the fallback paths are correct.

Reply:
Ah I understand thanks. I was struggling to follow the code - I played with it a bit, maybe some of these changes can help? Let me know what you think.

```go
func TestFavoriteLocations(t *testing.T) {
	favoriteLocations, _ := getFavoriteLocations()

	// This is a map containing the base name
	// of every key that should be in favoriteLocations
	expected := map[string]string{
		"Home":      "",
		"Documents": "Documents",
		"Downloads": "Downloads",
	}

	homeDir, err := os.UserHomeDir()
	assert.Nil(t, err)

	for name, subdir := range expected {
		fav, ok := favoriteLocations[name]
		if !ok {
			t.Errorf("missing favourite location", name)
			continue
		}

		if subdir == "" {
			assert.Equal(t, homeDir, fav)
		} else {
//			assert.Equal(t, homeDir, filepath.Dir(fav)) // I don't think this assertion must be true, the user could specify a different parent directory
			_, err := os.Stat(fav)
			if err == nil {
				continue // folder found
			}
			if !os.IsNotExist(err) {
				t.Errorf("failed to read directory %s", fav)
				continue
			}
			fallbackName := filepath.Base(fav)
			if !strings.EqualFold(subdir, fallbackName) {
				t.Errorf("%s should equal fold with %s", subdir, fallbackName)
			}
		}
	}
}
```

Reply:
So the only thing that's different here is that
```go
assert.Equal(t, homeDir, filepath.Dir(fav))
```
has been removed. Right?